[Scraper] Support for filename identifiers

sundermann opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->
Parse filenames for identifier tags and pass them through `getDetails` by providing an optional `uniqueIDs` parameter.

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
For quite some time now, [Jellyfin](https://jellyfin.org/docs/general/server/media/movies/), [Emby](https://emby.media/support/articles/Movie-Naming.html#id-tags-in-folder--file-names), and [Plex](https://support.plex.tv/articles/naming-and-organizing-your-tv-show-files/) have been supporting supplying database identifier tags inside the filename. The naming is not completely consistent between these three, the general idea of `dbidentifier-XXXX` is common amongst them. Kodi does not officially support any of these tagging conventions. What's worse though is that kodi fails a tagged movie like this completely. Assuming a Jellyfin like tagging for a movie like this
```
Avatar (2009) [tmdbid-19995]&title=Avatar (2009)/Avatar (2009).mkv
```
Kodi would parse the file and misinterpret `tmdbid` as year and the title is parsed wrong too:
```
title=Avatar (2009) [tmdbid
year=1999
```
This happens because the movie title is passed into `CUtil::CleanString` which apparently breaks here.

This change additionally parses the filename for such identifiers and passes them into `getDetails` using the `uniqueIDs` parameter.

There was already an [attempt](https://github.com/xbmc/metadata.themoviedb.org.python/pull/182) at fixing this at the addon level, however it required adding the identifier after the year. This also shows that there is a valid use case for this amongst users.

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
OLED77C28LB / webOS7/22
MacOS

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->
Add support for filename identifiers

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

sundermann committed:
[Scraper] Parse filename identifiers

This allows i.e. supporting filename database identifiers

sundermann commented:
@the-black-eagle You seem to have touched the scraper a bit lately, would you mind reviewing?

jjd-uk commented:
@rmrector looks after the TMDB scraper so perhaps he might be able to provide input.

hexoverflow commented:
Is there progress on this matter? It would be nice if kodi would finally be able to support database identifiers in filenames and I grasp that this PR is a prerequisite to support that.

sundermann commented:
Need to find someone to review. Pinging @rmrector if he would be so kind to review

garbear reviewed (approved):
Without much knowledge of the scraper system, I still have no problem leaving a +1 on what looks like a safe single line addition that does what the PR title says.

rmrector commented:
I don't think each individual scraper should be responsible for common things like patterns for finding DB identifiers in a filename.

> Kodi does not officially support any of these tagging conventions. What's worse though is that kodi fails a tagged movie like this completely.
> Kodi would parse the file and misinterpret tmdbid as year and the title is parsed wrong too:
> This happens because the movie title is passed into `CUtil::CleanString` which apparently breaks here.

A problem has been identified, but this seems to push the responsibility to the wrong place. I would suggest improving Kodi's filename handling.

sundermann commented:
@rmrector I agree. `CUtil::CleanString` should not break in this case. However, I still think that optional filename identifiers would be an improvement. What if we moved identifier matching into Kodi passing `database` and `identifier` to scrapers. Those could then try to lookup from the identifier directly rather than trying to match from title and year.

rmrector commented:
Ya, I like that idea. That should match the **uniqueid** system we already have - [see how they render in a movie NFO file on the wiki](https://kodi.wiki/view/NFO_files/Movies#nfo_Tags).

When you have these, you can then call the "getdetails" function of the scraper directly, rather than "find". [Python TV scrapers standardized on repurposing the "url" param](https://forum.kodi.tv/showthread.php?tid=370489) (or "&lt;episodeguide&gt;" in that forum post) with JSON-formatted unique IDs, as does the Python TMDB movie scraper, so you _might_ be able to use that, too. If you call the Python scraper "getdetails" function with a "url" with a value like `{"tmdb": "19995"}`, such scrapers should Just Work without changes to each scraper.

An alternative is to add an explicit parameter "uniqueid" to the "getdetails" function for this purpose and change scrapers to match. If you go that way, the same parameter will be useful to add to "getepisodelist", "getepisodedetails", and "getartwork" functions. The info for those will come from the DB or a previous scraper call, rather than file name, but it would be nice to scraper devs to make them available at the same time.

sundermann commented:
@rmrector I have changed `CUtil::CleanString` to properly parse filenames containing db identifiers and return that information. However, I'm not sure if I understand your idea correctly. Would that mean to skip the whole call chain to `FindMovie/Show` in case a db identifier is found?

sundermann commented:
@rmrector Not sure if skipping `FindMovie` is actually a good idea. The scraper should make sure the unique database identifier is actually supported and if not fall back to looking up via title and year. I guess it's still better to do the matching in kodi itself and pass a `database` and `identifier` property to the scrapers if found.

rmrector commented:
Yes, that was my suggestion. "Find" is search by title and year, but we're not searching. It would be similar to putting a URL in an NFO file or what Kodi does with MKV tags `IMDBURL` and `TMDBURL`.

If the direct lookup fails, _Kodi_ should fall back to search by title and year.

rmrector reviewed (changes requested):
Thanks for the submission and updates. I've added a couple of comments.

sundermann commented:
@rmrector Thanks for your feedback. The current approach works fine using the tmdb scraper for movies. However, I am having some trouble making it work for tv shows. I am testing with the tmdb shows scraper and it does not expect a json parameter. I tried setting the `<episodeguide>` tag inside a `IVideoInfoTagLoader` but didn't had much success doing so.

sundermann commented:
Reworked it again a bit. I noticed it is already possible to pass external ids in the form of `identifierType/identifier` to the TMDB movie and shows scrapers. In case of a filename identifier it tries to look up via filename identifier first and falls back to the default behaviour if no result was found or the scraper does not support searching for external ids. Even though it'd be nice to skip the find step, there's a problem here. Looking at different scraper implementatinos it seems there's no standard on what the url must be like. In case of the TMDB movie scraper everything is well as it excepts a json object containing the identifier type and identifier. However, the TMDB series scraper for instance needs the TMDB id inside. In case imdb filename identifiers are used this does not work.

rmrector commented:
The process I'm suggesting is that if we have one or more IDs, look up / "getdetails" by ID. If the lookup fails, try the original "find" by name. These are two separate processes that should be kept separate.

> Looking at different scraper implementations it seems there's no standard on what the url must be like.

Maybe the alternative is more compelling - passing the JSON object as a new `uniqueids` parameter to the "getdetails" call and adjust scrapers to match. That expands the scope of changes, but then we're explicit and can move away from the ambiguous url parameter.

sundermann commented:
@rmrector Thank you for your input. I have changed the code to not call `FindMovie` at all and instead call `GetDetails` directly. I have only implemented the movie case for now but the tv show case can be implemented in the same way. Can you check if this is going in the right direction?

Another thing to consider is that `GetDetails` does not fail gracefully when passed with an invalid ID. Also in that case `GetDetails` was not return false so I had to add an additional condition in `CPluginDirectory::GetPluginResult` though that seems wrong:
```
[metadata.themoviedb.org.python]: The Movie Database Python error with web service TMDB: failed to reach the remote site Reason: Not Found
```

sundermann commented:
`DetailsFromFileItem` does return now if information was parsed. The error handling needs to be fixed inside the scraper though.

sundermann commented:
Looking a bit into the tmdb API docs I'm not sure if calling `GetDetails` directly is actually correct. It seems to be a special case only for movies that the [Details](https://developer.themoviedb.org/reference/movie-details) endpoint can accept both a `tmdb` and a `imdb` identifier. Passing a `imdb` identifier to the tv show [Details](https://developer.themoviedb.org/reference/tv-series-details) endpoint it fails with an error. Passing an `imdb` identifier to `Details` is not mentioned on the API docs neither for movies nor tv shows. So in the movie case it actually works but it's not guaranteed to work by the API docs.

I would make the case that the scraper should call the [Find By ID](https://developer.themoviedb.org/reference/find-by-id) endpoint which accepts the `identifierType`together with an `identifier`. The only shortcut that could be made is if the scraper knows that the passed `identifierType` can be passed directly into `GetDetails`. Regarding the tmdb scrapers this is only the case for `tmdb` (and the out of spec `idmb` for movies only) identifiers. Any other identifier like i.e. `tvdb` has to go through `Find By ID` anyway.

sundermann commented:
@rmrector Thanks, I have added two new explicit parameters for `identifierType` and `identifier` to `getDetails` and extended the implementation to cover tv shows and music videos. I have also added two commits to showcase the implementation for TMDB and TMDB shows. I'll remove those commits later as they should target the specific repositories instead, but it's probably useful to see how this can be implemented on the scraper side for completeness of this PR.

hexoverflow commented:
I managed to compile the custom build along with the changes to the scrapers and it works fine for me. Finally I can enjoy organizing my collection using filebot with full support for filename identifiers. Any news though on when this will be merged?

sundermann commented on xbmc/addons/Scraper.cpp:
```diff
@@ -846,34 +847,54 @@ void DetailsFromFileItem<CArtist>(const CFileItem &item, CArtist &artist)
   std::stringstream str;
   str << "plugin://" << ID << "?action=" << action << "&" << key << "=" << CURL::Encode(url);
   str << "&pathSettings=" << CURL::Encode(pathSettings);
+  str << "&identifierType=" << identifierType << "&identifier=" << identifier;
```
I have made the implementation more generic by accepting a map of identifiers to `getDetails`. The json object passed can look like this `{'tmdb': '1234', 'imdb': 'tt1234'}`.

KarellenX commented:
I wonder if @pkscout is aware of this. He looks after TMDB TV Shows

pkscout commented:
This is the first I've heard of this, but I'm happy to look at a PR to deal with whatever problem is getting solved here (I think I followed the thread properly and see what we're trying to address, but I'm not 100% sure).  Please do the pull request against the Nexus branch the TMDb TV Shows.  Thanks.

KarellenX commented:
I'll ping @szsori at TVDB

KarellenX commented:
Oh, and let's not forget @romanvm for TV Maze scraper

sundermann commented:
@pkscout I don't think separate branches are needed. This PR just adds another way of looking up by id and it's optional for scrapers to implement. If using Nexus that path cannot be called so that's fine too

romanvm commented:
What is the new convention for online database IDs in file names so I could test my scraper?

sundermann commented:
@romanvm The [emby docs](https://emby.media/support/articles/Movie-Naming.html#id-tags-in-folder--file-names) pretty much lists all of the different conventions that are also supported with this PR.

romanvm commented:
Do I understand correctly that multiple ids in a filename are not supported? E.g:
`[tvdb=XXXX,tvmaze=YYYY]` or `[tvdb=XXXX&tvmaze=YYYY]`

Tested TVmaze scraper with the current master - it throws an error if [tvdb=XXXX] tag is present in the file because `url` parameter is missing in the call. However, Kodi falls back to regular flow starting with `find` so scraping is done despite of an error. In any case my scraper needs to be fixed.

sundermann commented:
Yes, there can only be one filename identifier, looking at Jellyfin, Plex and others they do only support one identifier. However, the scraper interface is already prepared to accept multiple ids like `uniqueIDs={"tmdb": "1234", "imdb": "4321"}`.

thezoggy commented:
also to note, not doing this at just filename but ideally movie/tvseries folder to help assist apps to know what movie the scraper should link to is ideal. especially helps when users use their own title vs what the scraper has (or if the name on the scraper has changed over time). plex/jellyfin/emby pretty much all support the same `<provider>-<id>` with `{}` or `[]` wrapping it.

example from plex (square brackets also work fine):
> If you are using the “Plex TV Series” agent, you can optionally include the TMDB or TVDB show ID in the folder name to improve matching. If you choose to do that, it must be inside curly braces: `ShowName (2020) {tvdb-123456}` or `ShowName (2020) {tmdb-123456}` where 123456 is the show ID.

sundermann commented:
@thezoggy This applies at folder level for tvseries. It applies to movies as well if the "Movies are in separate folders" scraper setting is enabled

Danielku15 commented:
@sundermann I'm cross posting here regarding my forum request https://forum.kodi.tv/showthread.php?tid=382449 I came to notice some possible design flaw in this implementation. You can read about my concerns in the forum post. Any chance you can take statement on what I might be missing? 

In my thinking: As Kodi almost always seem to do a "find" followed by a specific "getdetails", adding the uniqueIDs to "getdetails" is too late. We need it in the "find" to decide already which exact show to provide to Kodi.

Danielku15 commented:
Thanks for the further insights. In the meanwhile I've setup a full workspace to compile and debug Kodi. Seems one key problem in my scraper setup I wanted to use dashes which are not included in the regex. 

The `HasFilenameIdentifier` check then failed and caused the new direct loading paths to be skipped. 

Just reading through the code paths lead me on some wrong tracks. Actually debugging helped me out. 

I'd say it would be a good choice to add dashes to the regex of the IDs to cover APIs that use UUIDs or in my case compound IDs like "series-1234". But beside that things seem to be working fine and my findings were wrong 🙇‍♂️
